obstacle_stop_planner_refine/src/obstacle_stop_planner.cpp

```diff
@@ -42,35 +40,286 @@ namespace obstacle_stop_planner
   const SlowDownControlParameter & slow_down_param,
   const AdaptiveCruiseControlParameter & acc_param)
 : node_(node),
-  vehicle_info_(vehicle_info),
-  stop_param_(stop_param),
-  slow_down_param_(slow_down_param),
-  acc_param_(acc_param)
+  vehicle_info_(vehicle_info)
 {
+  updateParameters(stop_param, slow_down_param, acc_param);
+
   debug_ptr_ = std::make_shared<ObstacleStopPlannerDebugNode>(
     node_,
-    vehicle_info_.wheel_base +
-    vehicle_info_.front_overhang);
+    vehicle_info_.wheel_base_m_ +
+    vehicle_info_.front_overhang_m_);
 
   // Initializer
   acc_controller_ = std::make_unique<obstacle_stop_planner::AdaptiveCruiseController>(
     node_, vehicle_info_, acc_param_);
 }
 
-void ObstacleStopPlanner::obstaclePointcloudCallback(
-  const sensor_msgs::msg::PointCloud2::ConstSharedPtr input_msg)
+void ObstacleStopPlanner::updateParameters(
+  const StopControlParameter & stop_param,
+  const SlowDownControlParameter & slow_down_param,
+  const AdaptiveCruiseControlParameter & acc_param)
 {
-  obstacle_pointcloud_ptr_ = updatePointCloud(input_msg);
+  stop_param_ = stop_param;
+  slow_down_param_ = slow_down_param;
+  acc_param_ = acc_param;
+
+  {
+    const auto & i = vehicle_info_;
+
+    stop_param_.stop_margin += i.wheel_base_m_ + i.front_overhang_m_;
+    stop_param_.min_behavior_stop_margin +=
+      i.wheel_base_m_ + i.front_overhang_m_;
+    slow_down_param_.slow_down_margin += i.wheel_base_m_ + i.front_overhang_m_;
+    stop_param_.stop_search_radius = stop_param.step_length + std::hypot(
+      i.vehicle_width_m_ / 2.0 + stop_param.expand_stop_range,
+      i.vehicle_length_m_ / 2.0);
+    slow_down_param_.slow_down_search_radius = stop_param.step_length + std::hypot(
+      i.vehicle_width_m_ / 2.0 + slow_down_param.expand_slow_down_range,
+      i.vehicle_length_m_ / 2.0);
+  }
 }
 
+Output ObstacleStopPlanner::processTrajectory(const Input & input)
+{
+  // Search obstacles near vehicle
+  const auto obstacles = searchCandidateObstacle(input.input_trajectory, input.obstacle_pointcloud);
+
+  // Find collision point
+  const auto nearest_collision = findCollisionPoint(input.input_trajectory, obstacles);
+
+  Output output;
+
+  // Limit trajectory for adaptive cruise control
+  if (nearest_collision.has_value()) {
+    const auto acc_trajectory = adaptiveCruise(input, nearest_collision.value());
+
+    if (acc_trajectory.has_value()) {
+      // Set adaptive cruise trajectory as output
+      output.output_trajectory = acc_trajectory.value();
+    } else {
+      // Slow down if obstacle is in the detection area
+      const auto slowdown_trajectory = slowDown(input.input_trajectory, nearest_collision.value());
+      // Stop if obstacle is in the detection area
+      output.output_trajectory = obstacleStop(slowdown_trajectory, nearest_collision.value());
+    }
+  } else {
+    // Set input trajectory as output
+    output.output_trajectory = input.input_trajectory;
```

Comment:
How about using `early return`? It can reduce the nest.

```cpp
if (!nearest_collision.has_value()) {
    output.output_trajectory = input.input_trajectory;
    return;
}